state/sequence.go

```diff
@@ -38,3 +43,128 @@ func (s *State) sequence(name string) (int, error) {
 	}
 	return result.Counter, nil
 }
+
+// sequenceWithMin safely increments a database backed sequence,
+// allowing for a minimum value for the sequence to be specified. The
+// minimum value is used as an initial value for the first use of a
+// particular sequence. The minimum value will also cause a sequence
+// value to jump ahead if the minimum is provided that is higher than
+// the current sequence value.
+//
+// The data manipulated by `sequence` and `sequenceWithMin` is the
+// same. It is safe to mix the 2 methods for the same sequence.
+//
+// `sequence` is more efficient than `sequenceWithMin` and should be
+// preferred if there is no minimum value requirement.
+func (st *State) sequenceWithMin(name string, minVal int) (int, error) {
+	sequences, closer := st.getCollection(sequenceC)
+	defer closer()
+	updater := &dbSeqUpdater{
+		coll: sequences,
+		id:   st.docID(name),
+		name: name,
+	}
+	return updateSeqWithMin(updater, minVal)
+}
+
+// seqUpdater abstracts away the database operations required for
+// updating a sequence.
+type seqUpdater interface {
+	read() (int, error)
+	create(value int) (bool, error)
+	set(expected, next int) (bool, error)
+}
+
+const maxSeqRetries = 20
+
+// updateSeqWithMin implements the abstract logic for incrementing a
+// database backed sequence in a concurrency aware way.
+//
+// It is complicated because MongoDB's atomic update primitives don't
+// provide a way to upsert a counter while also providing an initial
+// value.  Instead, a number of database operations are used for each
+// sequence update, relying on the atomicity guarantees that MongoDB
+// offers. Optimistic database updates are attempted with retries when
+// contention is observed.
+func updateSeqWithMin(sequence seqUpdater, minVal int) (int, error) {
+	for try := 0; try < maxSeqRetries; try++ {
+		curVal, err := sequence.read()
+		if err != nil {
+			return -1, errors.Annotate(err, "could not read sequence")
+		}
+		if curVal == 0 {
+			// No sequence document exists, create one.
+			ok, err := sequence.create(minVal + 1)
+			if err != nil {
+				return -1, errors.Annotate(err, "could not create sequence")
```

Comment:
Is this really an error? Why not a retry?

Perhaps recording the semantics of the create method in the interface.